org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/format/LSPFormatFilesHandler.java

```diff
@@ -69,37 +78,50 @@ public Object execute(final ExecutionEvent event) throws ExecutionException {
 		return null;
 	}
 
-	protected void formatFile(final @NonNull IFile file, final IProgressMonitor monitor) {
+	protected Future<Boolean> formatFile(final @NonNull IFile file, final IProgressMonitor monitor) {
 		if (!file.exists() || !LanguageServersRegistry.getInstance().canUseLanguageServer(file))
-			return;
+			return CompletableFuture.completedFuture(false);
 
 		final var docProvider = getDocumentProvider(file);
 		try {
 			docProvider.connect(file);
-			final IDocument doc = docProvider.getDocument(file);
-			if (doc == null)
-				return;
-
-			monitor.setTaskName(NLS.bind(Messages.LSPFormatFilesHandler_FormattingFile, file.getFullPath()));
-			final VersionedFormatRequest formatRequest = formatter.versionedRequestFormatting(doc,
-					new TextSelection(0, 0));
-
-			List<? extends TextEdit> formatEdits = formatRequest.edits().get();
-			if (formatEdits.isEmpty())
-				return;
-
-			docProvider.aboutToChange(doc);
-			UI.getDisplay().syncExec(() -> formatter.applyEdits(doc, formatEdits, formatRequest.version()));
-			docProvider.changed(doc);
-			docProvider.saveDocument(monitor, file, doc, true);
-		} catch (InterruptedException ex) {
-			Thread.currentThread().interrupt();
-			LanguageServerPlugin.logError(ex);
-		} catch (final Exception ex) {
+		} catch (CoreException ex) {
 			LanguageServerPlugin.logError(ex);
-		} finally {
-			docProvider.disconnect(file);
+			return CompletableFuture.completedFuture(false);
 		}
+		final IDocument doc = docProvider.getDocument(file);
+		if (doc == null)
+			return CompletableFuture.completedFuture(false);
+
+		return LanguageServers.forDocument(doc) //
+				.withFilter(LSPFormatter::supportsFormatting) //
+				.computeFirst((lsw, ls) -> {
```

Comment:
The new API is intended to provide the LS to you on a dedicated dispatch thread so it is safe to make a direct call to the LS. You shouldn't do arbitrary computations inside one of these lambda blocks. It will tie up the dispatch thread so that nothing else can be sent to the LS...